tests/test_unsafe.py

```diff
@@ -0,0 +1,374 @@
+from typing import ContextManager
+
+import pytest
+import yaml
+from jinja2.ext import Extension
+from plumbum import local
+from plumbum.cmd import git
+
+from copier.cli import CopierApp
+from copier.errors import UserMessageError
+from copier.main import run_copy, run_update
+from copier.types import AnyByStrDict
+
+from .helpers import build_file_tree
+
+
+class JinjaExtension(Extension):
+    ...
+
+
+@pytest.mark.parametrize(
+    ("spec", "expected"),
+    [
+        (
+            {},
+            does_not_raise(),
+        ),
+        (
+            {"_tasks": []},
+            does_not_raise(),
+        ),
+        (
+            {"_tasks": ["touch task.txt"]},
+            pytest.raises(
+                UserMessageError, match="Template uses unsafe feature: tasks"
+            ),
+        ),
+        (
+            {"_migrations": []},
+            does_not_raise(),
+        ),
+        (
+            {
+                "_migrations": [
+                    {
+                        "version": "v1",
+                        "before": ["touch v1-before.txt"],
+                        "after": ["touch v1-after.txt"],
+                    }
+                ]
+            },
+            does_not_raise(),
+        ),
+        (
+            {"_jinja_extensions": []},
+            does_not_raise(),
+        ),
+        (
+            {"_jinja_extensions": ["tests.test_unsafe.JinjaExtension"]},
+            pytest.raises(
+                UserMessageError,
+                match="Template uses unsafe feature: jinja_extensions",
+            ),
+        ),
+        (
+            {
+                "_tasks": ["touch task.txt"],
+                "_jinja_extensions": ["tests.test_unsafe.JinjaExtension"],
+            },
+            pytest.raises(
+                UserMessageError,
+                match="Template uses unsafe features: jinja_extensions, tasks",
+            ),
+        ),
+    ],
+)
+@pytest.mark.parametrize("unsafe", [False, True])
+def test_copy(
+    tmp_path_factory: pytest.TempPathFactory,
+    unsafe: bool,
+    spec: AnyByStrDict,
+    expected: ContextManager[None],
+) -> None:
+    src, dst = map(tmp_path_factory.mktemp, ["src", "dst"])
+    build_file_tree({(src / "copier.yaml"): yaml.safe_dump(spec)})
+    with does_not_raise() if unsafe else expected:
+        run_copy(str(src), dst, unsafe=unsafe)
+
+
+@pytest.mark.parametrize("unsafe", [False, True])
+def test_copy_cli(
+    tmp_path_factory: pytest.TempPathFactory,
+    capsys: pytest.CaptureFixture[str],
+    unsafe: bool,
+) -> None:
+    src, dst = map(tmp_path_factory.mktemp, ["src", "dst"])
+    build_file_tree(
+        {(src / "copier.yaml"): yaml.safe_dump({"_tasks": ["touch task.txt"]})}
+    )
+    _, retcode = CopierApp.run(
+        ["copier", "copy", *(["--UNSAFE"] if unsafe else []), str(src), str(dst)],
+        exit=False,
+    )
+    if unsafe:
+        assert retcode == 0
+    else:
+        assert retcode == 1
```

Comment:
Could we have a special return code for this kind of failure?

Reply:
Good idea. I've added a new error class and catch it in the CLI code to exit with code 2 in this case.